### DISCUSSION TITLE
files for creating a new tool (instead of a deprecrated plugin)

### DISCUSSION THREAD

**Discussion by dmotxilla** [danny-avila/LibreChat#3740] [Help Wanted]
I need to create a plugin/tool. I see that plugins are deprecreated. Is there a doc about creating a tool?
I only see this doc, which seems about deprecated plugins: https://www.librechat.ai/docs/development/tools_and_plugins

What files do I need to create/modify in order to create a Tool?
The TavilySearch is a good example to mimic? (it is a tool, not a plugin?)

- api/app/clients/tools/structured/TavilySearchResults.js
the actual tool
- api/app/clients/tools/structured/specs/TavilySearchResults.spec.js  
testing/spec

all these files need to be updated with the new tool? or some are for tools and others are for plugins?
- api/app/clients/tools/util/handleTools.js
- api/app/clients/tools/index.js
- api/app/clients/tools/manifest.json

and configuration in:
- .env

---
TavilySearch has only one function (with several parameters). My tool has several functions. Do I need to create a new tool for each function, or I can combine all the functions into one tool? What is the best existing example to mimic (AzureAISearch.js, ChatTools.js, DALLE3.js, TavilySearchResults.js...)?

---
Also, how can I enable the tool in the configuration, instead of manually setting up the tool in the web console, Assistants -> Create Assistant -> Add Tool?

**Comment by danny-avila** [comment]
Please see https://github.com/danny-avila/LibreChat/issues/3607

Creating tools will actually remain very much the same, as those current tools you listed are going to be backwards compatible:

> AzureAISearch.js, DALLE3.js, TavilySearchResults.js

**Comment by dmotxilla** [comment]
I've copied `api/app/clients/tools/structured/TavilySearchResults.js` to `api/app/clients/tools/structured/Tavily2SearchResults.js`,
and replaced all keywords (`Tavily2SearchResults`, `tavily2_search_results_json`, `TAVILY2_API_KEY`)
and created the corresponindig entries in:
- api/app/clients/tools/util/handleTools.js
- api/app/clients/tools/index.js
- api/app/clients/tools/manifest.json
- .env

docker compose down, up.
however, when I click on Assistants -> Create Assistants -> Add Tools, I do not see the new "Tavily2 Search". Why is that?

then, in manifest.json, I've removed the entry about "Stable Diffusion",
docker compose down, up,
but "Stable Diffusion" stills appears in Assistants -> Create Assistants -> Add Tools.

What can be the issue?

**Reply by danny-avila** [reply]
You need to build the image yourself, are you sure you're not pulling from the ghcr repository?

**Reply by dmotxilla** [reply]
yes, this must be the problem. I did: 
```
$ git clone https://github.com/danny-avila/LibreChat.git
$ cd LibreChat
# edit .env, librechat.yaml, docker-compose.yml, docker-compose.override.yml
# change the source files with the new tool
$ docker compose up -d
```
what do I need to do so to build the image myself and use it? like this? something else?
```
$ docker build -t ghcr.io/danny-avila/librechat-dev:david1 .
# edit docker-compose.yml, replace ghcr.io/danny-avila/librechat-dev:latest by ghcr.io/danny-avila/librechat-dev:david1?
$ docker compose up -d
```

**Reply by danny-avila** [reply]
You need to make use of a docker override file: https://www.librechat.ai/docs/configuration/docker_override

```yaml
services:
  api:
    volumes:
      - type: bind
        source: ./librechat.yaml
        target: /app/librechat.yaml
# Below is what's needed for building locally
    image: librechat
    build:
      context: .
      target: node
```

Then:

```bash
docker compose down
docker compose build --no-cache
docker compose up -d
```

**Reply by dmotxilla** [reply]
it works, thanks!

**Reply by danny-avila** [reply]
Sweet! For development, I recommend not using docker though: https://www.librechat.ai/docs/development/get_started

If containers/docker are a must, I have a codespaces branch will push up soon
